Automattic/wp-calypso
# Update Connect Domain screens to allow for Domain Connect

We want to use [Domain Connect](https://github.com/Domain-Connect/spec) for domains that support it and it would be best if we implement that in the "Connect domain" flow.

## Proposed Changes

* I've added new flow for Domain Connect with 2 steps - DC_START and DC_VERIFYING
* Stop using query param `step` for internal links - it would be more consistent to determine the last `connection_mode` from the mappingSetupInfo API result
* Inroduce loading placeholders while the `mappingSetupInfo` loads
* New START screen for Domain Connect that would redirect the customer to the Domain Connect signed URL
* Updated the links at the bottom (ConnectDomainStepSwitchSetupInfoLink) to show the new flow (simple) when Domain Connect is available
* Updated the backend to return two new properties in the `mappingSetupInfo` API that will contain hte connection_mode and the Domain Connect apply URL and also allow the `return_url` param to be encoded - D101993-code

## Preview of the new flow

### If Domain Connect is available you'll see this screen
![Screenshot 2023-02-28 at 21 44 37](https://user-images.githubusercontent.com/1355045/221963144-8111d8e7-e7a4-417d-95ab-9687959b2727.png)

### When you click on Start seetup you'll be sent to the DNS Provider
Once you log in you'll see this (or similar - that part depends on the DNS Provider):
![Screenshot 2023-02-28 at 21 47 25](https://user-images.githubusercontent.com/1355045/221963722-f69ee5f2-29ad-4ab9-9921-9319ee88395b.png)

### If you click on Connect you should get redirected back to this
![Screenshot 2023-02-28 at 21 43 55](https://user-images.githubusercontent.com/1355045/221963786-7e38b124-60a1-43f3-a332-6090585118c1.png)

### If you click on Cancel you'll get here
![Screenshot 2023-02-28 at 21 42 08](https://user-images.githubusercontent.com/1355045/221963898-520b7f10-26c7-4206-949b-afb3035d128e.png)

### If there's an error you'll see this
![Screenshot 2023-02-28 at 21 43 02](https://user-images.githubusercontent.com/1355045/221964007-8c91c019-b8ef-4a72-a32f-067fde5fb6e2.png)



## Testing Instructions
* Apply D101993-code and then build Calypso with this PR applied
* Make sure that mapping a domain that does not support Domain Connect still works as before (there is suggested and advanced flows)
* Then try to map a domain that supports Domain Connect - the initial step should now render the new START component
* The button in that component should show the DNS provider Domain Connect flow where you should be able to apply the provided template records with a single click
* Make sure you can click on the links to switch between the different flows (simple, manual, advanced)
* See what happens when you successfully apply the records at the DNS provider
* NOTE: you'll always be redirected back to wordpress.com. To test the flows properly just replace https://wordpress.com/ with http://calypso.localhost:3000/ but keep the rest of the URL

## Pre-merge Checklist
